### PR TITLE
rbd: check volume details from original volumeID

### LINKED ISSUE #2929: Volume mounts for Pods failing with RBD mirroring enabled
# Describe the bug #

Two ceph cluster deployed under same stretched kubernetes cluster managed through single rook-ceph operator.  Each ceph cluster is created and confined under respective namespaces.
After RBD mirroring is enabled between ceph clusters,  I am able to create PVCs/PV and even volumeattachments is there but when I spin up a pod trying to use that PVC, scheduling fails with following error:
`MountVolume.MountDevice failed for volume "pvc-3b696002-f3c2-49d6-a9fb-5bcbcc82b921" : rpc error: code = Internal desc = failed to establish the connection: failed to get connection: connecting failed: rados: ret=-13, Permission denied`

Based on investigation, it was determined that since same rook operator is used to create both clusters, ceph-csi config generated by operator and fed to `csi-rbdplugin` containing information for both clusters:
```
csi-cluster-config-json: '[{"clusterID":"rook-ceph-dc1","monitors":["10.3.5.151:6789","10.3.130.59:6789","10.3.53.161:6789"]},{"clusterID":"rook-ceph-dc2","monitors":["10.3.44.248:6789","10.3.168.71:6789","10.3.224.29:6789"]}]'
```
this most likely leads to rbdplugin trying to mount image through second ceph cluster instead of local cluster (the one targeted under storage class and for respective PVC).

relevant debug:
`cluster_mapping.go:157] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 found new clusterID mapping "rook-ceph-dc2" for existing clusterID "rook-ceph-dc1"`

As credentials set under storage class point to local ceph cluster and within the namespace, permission denied occurs as second ceph cluster does not accept set credentials. This was verified with the scenario of creating same credentials on both clusters and used that under block pool storage class. Permission denied error does not occur anymore however volume mount still fails as it tries to mount the image which is set to non-primary state in second cluster:

`Warning  FailedMount             6s    kubelet                  MountVolume.MountDevice failed for volume "pvc-7442f51a-481a-4788-a390-f6740df59fb5" : rpc error: code = Internal desc = rbd image ceph-blockpool/csi-vol-1911755f-9fae-11ec-ba09-4641bc4d3bf8 is still being used`

This issue does not occur before the mirroring is enabled between ceph clusters. Without mirroring, Volume mounting against PVCs  created under each ceph clusters works.

# Environment details #

- Image/version of Ceph CSI driver : 3.5.1
- Helm chart version : rook/ceph:v1.8.6
- Kernel version :  5.15.17-200.fc35.x86_64
- Mounter used for mounting PVC (for cephFS its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) : krbd
- Kubernetes cluster version : 1.22.7
- Ceph cluster version : 16.2.7

# Steps to reproduce #

1. Setup details: Enable RBD mirroring. [Example setup](https://www.rook.io/docs/rook/v1.8/rbd-mirroring.html)
2. Deployment to trigger the issue: create PVC with using storage class configured for one of the ceph clusters
 ```
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rbd-pvc
  namespace: default
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: ceph-block-dc1
EOF
```
Spin up a sample pod using defined PVC

# Actual results #

Volume mount for pod fails with:
`MountVolume.MountDevice failed for volume "pvc-3b696002-f3c2-49d6-a9fb-5bcbcc82b921" : rpc error: code = Internal desc = failed to establish the connection: failed to get connection: connecting failed: rados: ret=-13, Permission denied`

# Expected behavior #

Volume mount for pod is done on target node and pod scheduling process is completed i.e container started and pod going into ready state.

# Logs #

If the issue is in PVC creation, deletion, cloning please attach complete logs
of below containers.

- csi-provisioner and csi-rbdplugin/csi-cephfsplugin container logs from the
  provisioner pod.
```
I0309 12:29:00.976190   60487 utils.go:191] ID: 82 GRPC call: /csi.v1.Node/NodeGetCapabilities
I0309 12:29:00.976310   60487 utils.go:195] ID: 82 GRPC request: {}
I0309 12:29:00.976577   60487 utils.go:202] ID: 82 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":3}}},{"Type":{"Rpc":{"type":5}}}]}
I0309 12:29:00.996152   60487 utils.go:191] ID: 83 GRPC call: /csi.v1.Node/NodeGetCapabilities
I0309 12:29:00.996293   60487 utils.go:195] ID: 83 GRPC request: {}
I0309 12:29:00.996485   60487 utils.go:202] ID: 83 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":3}}},{"Type":{"Rpc":{"type":5}}}]}
I0309 12:29:00.998092   60487 utils.go:191] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 GRPC call: /csi.v1.Node/NodeStageVolume
I0309 12:29:00.998401   60487 utils.go:195] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-2f6447ee-7355-42ca-9f74-d321adad56c2/globalmount","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":7}},"volume_context":{"clusterID":"rook-ceph-dc1","csi.storage.k8s.io/pv/name":"pvc-2f6447ee-7355-42ca-9f74-d321adad56c2","csi.storage.k8s.io/pvc/name":"rbd-pvc-mirror","csi.storage.k8s.io/pvc/namespace":"default","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4","journalPool":"ceph-blockpool","pool":"ceph-blockpool","storage.kubernetes.io/csiProvisionerIdentity":"1646825016236-8081-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4"}
I0309 12:29:00.998959   60487 cluster_mapping.go:157] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 found new clusterID mapping "rook-ceph-dc2" for existing clusterID "rook-ceph-dc1"
I0309 12:29:00.999150   60487 rbd_util.go:1296] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 setting disableInUseChecks: false image features: [layering] mounter: rbd
E0309 12:29:01.029990   60487 nodeserver.go:203] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 failed to establish cluster connection: failed to establish the connection: failed to get connection: connecting failed: rados: ret=-13, Permission denied
E0309 12:29:01.030164   60487 utils.go:200] ID: 84 Req-ID: 0001-000d-rook-ceph-dc1-0000000000000002-a0e006b5-9fa0-11ec-a0e8-9e12c024f7a4 GRPC error: rpc error: code = Internal desc = failed to establish the connection: failed to get connection: connecting failed: rados: ret=-13, Permission denied
```

# Additional context #

@Madhu-1 Already has a fix for this issue where cluster ID mapping is corrected. Tested csphcsi image with fix include and expected result was achieved.

### PR DESCRIPTION
**Pull request by Madhu-1**
Checking volume details for the existing volumeID first. if details like OMAP, RBD Image, Pool does not exist try to use clusterIDMapping to look for the correct pieces of information.

fixes: #2929

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Facing some CI issue and looks like its due to the rbd issue https://tracker.ceph.com/issues/54593

### CODE REVIEW

**Comment by Madhu-1** [comment]
> /retest all

something wrong with CI restarting all tests

**Review by Rakshith-R — DISMISSED** [review]

**Comment by humblec** [comment]
hmmm.. just wondering how it react to migrated volume handle.

**Comment by Madhu-1** [comment]
> hmmm.. just wondering how it react to migrated volume handle.

@humblec what is the concern here?

**Comment by humblec** [comment]
> > hmmm.. just wondering how it react to migrated volume handle.
> 
> @humblec what is the concern here?

looking some more into this, ideally we should be good, that said the static volume ID is untouched and should folow the code path as before, for new volume Handles only the change in effect . @Madhu-1 can you please double confirm ?

**Comment by Madhu-1** [comment]
> > > hmmm.. just wondering how it react to migrated volume handle.
> > 
> > 
> > @humblec what is the concern here?
> 
> looking some more into this, ideally we should be good, that said the static volume ID is untouched and should folow the code path as before, for new volume Handles only the change in effect . @Madhu-1 can you please double confirm ?

Yes this is not for static volume its for dynamic volume already getting called in DeleteVolume Code https://github.com/ceph/ceph-csi/blob/c3e35f88493ddd626f6818a48f30900d8166f4e2/internal/rbd/controllerserver.go#L762-L766

**Comment by Madhu-1** [comment]
`[](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.23/detail/mini-e2e-helm_k8s-1.23/863/pipeline/#step-102-log-789)[](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.23/detail/mini-e2e-helm_k8s-1.23/863/pipeline/#step-102-log-790)Mar 14 13:27:18.892: backend images not matching kubernetes resource count,image count 22 kubernetes resource count 21

    backend image Info:

     [csi-vol-61a6638c-a39a-11ec-b84e-46359091d2ce csi-vol-637f4854-a39a-11ec-b84e-46359091d2ce csi-vol-637f4854-a39a-11ec-b84e-46359091d2ce-temp csi-vol-63bdd485-a39a-11ec-b84e-46359091d2ce-temp csi-vol-63db0526-a39a-11ec-b84e-46359091d2ce csi-vol-63db0526-a39a-11ec-b84e-46359091d2ce-temp csi-vol-63f96934-a39a-11ec-b84e-46359091d2ce csi-vol-63f96934-a39a-11ec-b84e-46359091d2ce-temp csi-vol-643f9696-a39a-11ec-b84e-46359091d2ce csi-vol-643f9696-a39a-11ec-b84e-46359091d2ce-temp csi-vol-6457b69e-a39a-11ec-b84e-46359091d2ce csi-vol-6457b69e-a39a-11ec-b84e-46359091d2ce-temp csi-vol-6475acf7-a39a-11ec-b84e-46359091d2ce csi-vol-6475acf7-a39a-11ec-b84e-46359091d2ce-temp csi-vol-6488cc64-a39a-11ec-b84e-46359091d2ce csi-vol-6488cc64-a39a-11ec-b84e-46359091d2ce-temp csi-vol-64c087be-a39a-11ec-b84e-46359091d2ce csi-vol-64c087be-a39a-11ec-b84e-46359091d2ce-temp csi-vol-681b9d97-a39a-11ec-b84e-46359091d2ce csi-vol-681b9d97-a39a-11ec-b84e-46359091d2ce-temp csi-vol-6876ee6d-a39a-11ec-b84e-46359091d2ce csi-vol-6876ee6d-a39a-11ec-b84e-46359091d2ce-temp
`
Failed due to known errors.

**Review by humblec — DISMISSED** [review]

**Comment by Rakshith-R** [comment]
@mregifyio rebase

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
@mergifyio rebase

**Review by Rakshith-R — DISMISSED** [review]

**Comment by Madhu-1** [comment]
>E0322 05:03:30.732970  150121 utils.go:200] ID: 110 Req-ID: 0001-0024-4f124b93-57df-4ad4-a74c-e79193bc0971-0000000000000007-64d7469e-a99d-11ec-98e9-7af2c48ecac9 GRPC error: rpc error: code = Internal desc = error generating volume (0001-0024-4f124b93-57df-4ad4-a74c-e79193bc0971-0000000000000007-64d7469e-a99d-11ec-98e9-7af2c48ecac9): pool not found: pool ID(7) not found in Ceph cluster

frequently hitting this one. looks like its due to the rbd issue https://tracker.ceph.com/issues/54593

File: internal/rbd/nodeserver.go [hunk 1]
```diff
@@ -187,33 +185,13 @@ func populateRbdVol(
...
-			return nil, status.Error(codes.Internal, err.Error())
-		}
-		defer j.Destroy()
-
-		imageAttributes, err = j.GetImageAttributes(
-			ctx, rv.Pool, vi.ObjectUUID, false)
-		if err != nil {
-			err = fmt.Errorf("error fetching image attributes for volume ID (%s): %w", volID, err)
+			rv.Destroy()
+			err = fmt.Errorf("error generating volume (%s): %w", volID, err)
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: why assign `err` and not use `status.Errorf()` below?

**Review by nixpanic — COMMENTED** [review]
I think this should be fine. Not sure why @humblec is concerned about migrated handles, there is no comment or special testing for handles here, at least as far as I can see.

File: internal/rbd/nodeserver.go (hunk 1 above)
Comment (Madhu-1) [reply internal/rbd/nodeserver.go]: Yes, that can be done, Updated to use status.Errorf

**Review by Rakshith-R — DISMISSED** [review]

**Review by humblec — DISMISSED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Madhu-1 may be rebase ?

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
Common failure:
```
    Apr  4 07:53:17.339: failed to validate clones in different pool: creating PVCs and applications failed, 10 errors were logged
```

**Comment by Madhu-1** [comment]
> frequently hitting this one. looks like its due to the rbd issue https://tracker.ceph.com/issues/54593

frequently hitting this one. looks like its due to the rbd issue https://tracker.ceph.com/issues/54593 , will rebase once we update the base image with quincy.

**Comment by nixpanic** [comment]
> > frequently hitting this one. looks like its due to the rbd issue https://tracker.ceph.com/issues/54593
> 
> frequently hitting this one. looks like its due to the rbd issue https://tracker.ceph.com/issues/54593 , will rebase once we update the base image with quincy.

Ok, backport for Pacific is pending too ceph/ceph#45586. Should we track this as an issue?

**Comment by humblec** [comment]
Lets give one more try with Quincy . Interestingly the CI failure is only happening here in this PR..

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by Madhu-1** [comment]
Moving it out of 3.6 release as need to check what is going wrong.

**Comment by humblec** [comment]
> Moving it out of 3.6 release as need to check what is going wrong.

Thanks @Madhu-1 even I was thinkig the same

**Comment by humblec** [comment]
Cutting release branch... :+1:

**Review by nixpanic — CHANGES_REQUESTED** [review]
After dropping the commit to update the base image, it should be good to go?

@humblec what do you think?

**Comment by Madhu-1** [comment]
@nixpanic this fix is still not available in the ceph release. will revisit this one later.

**Comment by nixpanic** [comment]
> @nixpanic this fix is still not available in the ceph release. will revisit this one later.

Please link the Ceph issue tracker that this depends on, that makes it a little easier to review.

**Comment by nixpanic** [comment]
> @nixpanic this fix is still not available in the ceph release. will revisit this one later.

Marking as **Draft** for the time being.

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Review by Rakshith-R — DISMISSED** [review]

**Comment by Madhu-1** [comment]
@mergifyio rebase

File: internal/rbd/nodeserver.go [hunk 2]
```diff
@@ -189,33 +186,13 @@ func populateRbdVol(
 		}
 		rv.RbdImageName = volID
 	} else {
-		var vi util.CSIIdentifier
-		var imageAttributes *journal.ImageAttributes
-		err = vi.DecomposeCSIID(volID)
+		rv, err = GenVolFromVolID(ctx, volID, cr, req.GetSecrets())
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go:189]: `rv` was just created at line 174. Why is it needed to create it again (and potentially leak something)?

Comment (Madhu-1) [reply internal/rbd/nodeserver.go:189]: We need to reassign to `rv` after finding the right rbdVolume from the clusterID mapping. More details are provided in the issue https://github.com/ceph/ceph-csi/issues/2929#issue-1165094633

Comment (Madhu-1) [reply internal/rbd/nodeserver.go:189]: Let me see if i can rearrange the code block, its been many months.

Comment (nixpanic) [reply internal/rbd/nodeserver.go:189]: That is not clear from the code. Please add comments that explain this. Or, assign `rv` only once, move line 174 inside the `if isStaticVol`?

Comment (Madhu-1) [reply internal/rbd/nodeserver.go:189]: Code is rearranged now, PTAL

File: internal/rbd/nodeserver.go [hunk 3]
```diff
@@ -173,12 +170,7 @@ func populateRbdVol(
 
 		disableInUseChecks = true
 	}
-
-	rv, err := genVolFromVolumeOptions(ctx, req.GetVolumeContext(), disableInUseChecks, true)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-
+	rv := &rbdVolume{}
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: **Nit:** This works, but creating an empty object before overwriting it is a little ugly.

I prefer to use this instead:
```go
var rv *rbdVolume
```

File: internal/rbd/nodeserver.go (hunk 2 above)
Comment (nixpanic) [reply internal/rbd/nodeserver.go:189]: thanks!

**Review by nixpanic — DISMISSED** [review]

**Review by nixpanic — APPROVED** [review]

**Comment by nixpanic** [comment]
@mergifyio rebase

**Review by Rakshith-R — APPROVED** [review]

**Pull request merged**